WordPress/gutenberg
# Makes cover block dynamic and adds featured image binding

## What?
Closes #13795 and advances #37753.

Alternative to #39410.

- [x] Known bug: sometimes the featured image is not set when toggled.

## Why?
This approach doesn't care about keeping cover agnostic from WordPress and copies the method used by the `FeaturedImage` block to bind the block's background to the featured image.

## How?
We get the featured image from the nearest entity. If the block is not in a post, or does not get the postId from context, the binding (including UI) is unavailable. Once the binding is active - via setting a new attribute `useFeaturedImage` - then we watch for updates outside of the block to the post featured image.

To handle the fact that the featured image can be changed else where we changed the `save` function of the cover block to replace the url on the fly with a token `WordPress://featured-image` which gets replaced with the live featured image by the new cover block's server side renderer.

~~This `WordPress://featured-image` is not a perfect solution, the other option would be, but not sure if possible, to not run any save (set save to `null`) and reimplement the rendering on the server side.~~ Replaced with server side `preg_replace` injection of HTML.

## Testing Instructions
- switch to this branch
- create a post
- add a cover block, and pick an image from the media library
- add a featured image to the post
- on the cover block click the featured image icon in the toolbar
= the featured image should appear instead
- change the post's featured image
= the cover block's background should change
- install a plugin that can change the featured image (e.g. Featured Image by URL)
- change the post's featured image via the plugin
= verify that on the front end the cover block renders with the image you've set.

## Screenshots or screencast 

![dynamic-cover-block](https://user-images.githubusercontent.com/107534/159577603-ae230019-cfd5-4d01-955c-076773a0d805.gif)
